docker/Dockerfile

```diff
@@ -256,6 +271,47 @@ RUN set -e; \
     /etc/roaster/scripts/setup.sh fpm ccache; \
     truncate -s0 ~/.bash_history;
 
+RUN set -e; \
```

Comment:
If it's an arm-only later, why even bother to canonicalize x86?
To me it should be as simple as `if uname -m | grep -e'^'{aarch64,arm64}'$' >/dev/null; then`

Reply:
canonicalizing may not be a good design.
You're basically "tagging" uname output.
My experience tags won't keep mutual exclusive, you'll eventually start to check for things like 32/64 bit, or distinguish between amd64/x86_64.
Let's why I prefer simple check like this.

Reply:
> If it's an arm-only later, why even bother to canonicalize x86?

If we have to have it, why not adding support for x86? If something wrong with the Bazel apt source then at least we can retry it here.

> canonicalizing may not be a good design.
You're basically "tagging" uname output.
My experience is, tags won't keep mutual exclusive, you'll eventually start to check for things like 32/64 bit, or distinguish between amd64/x86_64.
Let's why I prefer simple check like this.

At least in this case, the architecture name is mutual exclusive. You have to decide which package to download and it has to be exactly one package. If we somehow need to check 32/64 bits (unlikely), or distinguish x86_64/amd64 (I don't know what is the difference, I think in the most cases they mean the same thing and people just pick one to represent all of x86_64/x64/amd64/k8 things), we can change the case stmt for that specific usecase (maybe even a different varaible name).

Canonicalization is a good design at least for the most of the cases, when you don't care about the differences between AMD/Intel CPUs or different generations. If you do need to distinguish, we may need to use more reliable ways (maybe read cpuinfo, try some special asm instructions or so, I am not sure, but definitely not relying on the wording of amd64/x86_64/k8).

Reply:
> If something wrong with the Bazel apt source then at least we can retry it here.

topological failure should not be retried, that will add more trouble than what it solved.

> I don't know what is the difference

- amd64 is AMD's version
- x86_64 is Intel's version, could be different from AMD's version on ISA extensions or vendor-specific things.
- x64 is Microsoft's name
- k8 is micro arch, not ISA, and I've never only seen it used by google sometimes.

> Canonicalization is a good design at least for the most of the cases

No, it's just a grouping of equivalent classes. It's only good if there's exactly one reasonable way of defining equivalence.
In this case, it's not, so it's not a good abstraction.
This doesn't mean it cannot be a good implementation as it can solve our problems, but don't try to abstract it by preparing for future extensions.

> when you don't care about the differences between AMD/Intel CPUs or different generations

Those are just examples I used to show you don't have exactly one set of bases.

Reply:
> amd64 is AMD's version
> x86_64 is Intel's version, could be different from AMD's version on ISA extensions or vendor-specific things.
> x64 is Microsoft's name
> k8 is micro arch, not ISA, and I've never only seen it used by google sometimes.

All correct, but these are etymologic differences, not how they are used for.

- debian/ubuntu apt sources uses `amd64`, no matter if it is Intel or AMD
- Rosetta 2 gives you x86_64, and obviously it is not Intel
- When you download packages, they won't distinguish if your x86 CPU is Intel or AMD. In our case, we download "bazel x86_64" and obviously it can be used on AMD CPUs
- Bazel build uses k8 to represent x86_64 builds

So my point is, you cannot rely on the `uname -m` wording to decuct anything more than "it is a 64-bit x86 CPU", then why make this problem over-complicated?

Reply:
Also, on M1 Mac, `uname -m` gives you arm64 while the docker containers running on it gives you `aarch64`. So `uname -m` is basically picking up random names from the equivalent group, that is why I prefer the canonicalized `TARGETARCH`. If there are not really any standard on this varaible either, then I prefer to make my own.

Reply:
Anyway I don't think it is a meaningful topic to discuss more on.

Reply:
> debian/ubuntu apt sources uses amd64, no matter if it is Intel or AMD

Because for x86/x87/SSE, amd64 and x86_64 are mostly compatible, and let's where most OS stops.

> Rosetta 2 gives you x86_64, and obviously it is not Intel

What do you mean? Rosetta transcode intel binary to arm.

> When you download packages, they won't distinguish if your x86 CPU is Intel or AMD. In our case, we download "bazel x86_64" and obviously it can be used on AMD CPUs

If you only talks about x86, mostly right. If you go beyond to ext, no.

> Bazel build uses k8 to represent x86_64 builds

Yes that's why I say google is the only one doing so. I guess they're using k8 for AMD K8, the first 64bit x86-compatible micro arch.

> Also, on M1 Mac, uname -m gives you arm64 while the docker containers running on it gives you aarch64. So uname -m is basically picking up random names from the equivalent group

No, BSD vs. Linux.

Reply:
We have deviated too far. We are discussing why we should/should not canonicalize the arch name here. The history of different names or the subtle difference of those names are irrelevant here.

Reply:
Or we are not even discussing canonicalization, but if should bring in x86 cases or `if uname -m | grep -e'^'{aarch64,arm64}'$' >/dev/null; then`

I prefer the former because it won't make people think "what about x86" when looking at this layer. Also if anyone wants to play yet another architecture (Unlikely because currently bazel only gives you x86_64 and arm64 prebuilts), they may reuse the arch selection case stmt here.